email-reader/src/main/java/no/sikt/nva/email/reader/handler/ScopusEmailReader.java

```diff
@@ -4,46 +4,95 @@
+import io.vavr.control.Try;
 import no.sikt.nva.email.reader.mapper.messagebodyreader.EmailParser;
 import no.sikt.nva.email.reader.mapper.messagebodyreader.MultipartReader;
 import no.sikt.nva.email.reader.mapper.messagebodyreader.ScopusEmailValidator;
+import no.sikt.nva.email.reader.model.exception.EmailException;
+import no.sikt.nva.email.reader.service.FileRetriever;
+import no.sikt.nva.email.reader.service.ZipFileRetriever;
 import no.unit.nva.s3.S3Driver;
+import nva.commons.core.Environment;
+import nva.commons.core.JacocoGenerated;
+import nva.commons.core.paths.UnixPath;
 import nva.commons.core.paths.UriWrapper;
 import org.apache.james.mime4j.dom.Message;
 import software.amazon.awssdk.services.s3.S3Client;
-import nva.commons.core.JacocoGenerated;
 
+import java.io.InputStream;
 import java.net.URI;
 import java.net.http.HttpClient;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 
 public class ScopusEmailReader implements RequestHandler<S3Event, Set<URI>> {
 
 
+    public static final String UNABLE_TO_DOWNLOAD_FILE = "Unable to download file";
+    public static final String COULD_NOT_PERSIST_FILE_IN_S_3_BUCKET = "Could not persist file in s3 bucket";
     private static final String S3_URI_TEMPLATE = "s3://%s/%s";
     private static final int SINGLE_EXPECTED_RECORD = 0;
     private final S3Client s3Client;
 
-    @SuppressWarnings({"PMD.UnusedPrivateField", "PMD.SingularField"})
-    private final HttpClient httpClient; //TODO: implement downloading
+    private final FileRetriever fileRetriever;
+    private final String scopusZipBucket;
 
     @JacocoGenerated
     public ScopusEmailReader() {
         this(S3Driver.defaultS3Client().build(),
-                HttpClient.newBuilder().build());
+                new ZipFileRetriever(HttpClient.newBuilder().build()),
+                new Environment().readEnv("SCOPUS_ZIP_BUCKET"));
     }
 
-    public ScopusEmailReader(S3Client s3Client, HttpClient httpClient) {
+    public ScopusEmailReader(S3Client s3Client, FileRetriever fileRetriever, String scopusZipBucket) {
         this.s3Client = s3Client;
-        this.httpClient = httpClient;
+        this.fileRetriever = fileRetriever;
+        this.scopusZipBucket = scopusZipBucket;
     }
 
     @Override
     public Set<URI> handleRequest(S3Event event, Context context) {
-        var emailString = getEmailFromS3(event);
-        var message = extractMessage(event, emailString);
-        return extractUrisFromMessage(event, message);
+        return Try.of(() -> getEmailFromS3(event))
+                .mapTry(email -> extractMessage(event, email))
+                .mapTry(message -> extractUrisFromMessage(event, message))
+                .mapTry(uriSet -> downloadZipFilesAndPersistThem(uriSet, event))
+                .getOrElseThrow(throwable -> handleFailure(throwable, event));
+    }
+
+    private Set<URI> downloadZipFilesAndPersistThem(Set<URI> uris, S3Event event) {
```

Comment:
Honestly, in normal parlance, download implies persistence, but implementation details, so perhaps `downloadToBucketStorage`. I don't think ZIP is important here, you're not zipping/unzipping.

Reply:
Fixed